src/PoissonSolvers/LaplaceHelpers.h

```diff
@@ -0,0 +1,389 @@
+                    const bool not_left_boundary = (global_index != 0);
+                    const bool right_boundary    = (global_index == size - 1);
+
+                    coords[d] -= 1;
+                    auto&& left = apply(u_m, coords);
+
+                    coords[d] += 2;
+                    auto&& right = apply(u_m, coords);
+
+                    res += hvector_m[d] * (not_left_boundary * left + right_boundary * right);
+                }
+                return res;
+            }
+
+        private:
+            using Mesh_t      = typename E::Mesh_t;
+            using Layout_t    = typename E::Layout_t;
+            using vector_type = typename Mesh_t::vector_type;
+            using domain_type = typename Layout_t::NDIndex_t;
+            const E u_m;
+            const vector_type hvector_m;
+            const unsigned nghosts_m;
+            const domain_type ldom_m;
+            const domain_type domain_m;
+        };
+
+        template <typename E>
+        struct meta_upper_laplace
+            : public Expression<meta_upper_laplace<E>,
+                                sizeof(E) + sizeof(typename E::Mesh_t::vector_type)
+                                    + 2 * sizeof(typename E::Layout_t::NDIndex_t)
+                                    + sizeof(unsigned)> {
+            constexpr static unsigned dim = E::dim;
+
+            KOKKOS_FUNCTION
+            meta_upper_laplace(const E& u, const typename E::Mesh_t::vector_type& hvector,
+                               unsigned nghosts, const typename E::Layout_t::NDIndex_t& ldom,
+                               const typename E::Layout_t::NDIndex_t& domain)
+                : u_m(u)
+                , hvector_m(hvector)
+                , nghosts_m(nghosts)
+                , ldom_m(ldom)
+                , domain_m(domain) {}
+
+            /*
+             * n-dimensional upper triangular Laplacian
+             */
+            template <typename... Idx>
+            KOKKOS_INLINE_FUNCTION auto operator()(const Idx... args) const {
+                using index_type = std::tuple_element_t<0, std::tuple<Idx...>>;
+                using T          = typename E::Mesh_t::value_type;
+                T res            = 0;
+
+                for (unsigned d = 0; d < dim; d++) {
+                    index_type coords[dim]        = {args...};
+                    const int global_index        = coords[d] + ldom_m[d].first() - nghosts_m;
+                    const int size                = domain_m.length()[d];
+                    const bool left_boundary      = (global_index == 0);
+                    const bool not_right_boundary = (global_index != size - 1);
+
+                    coords[d] -= 1;
+                    auto&& left = apply(u_m, coords);
+
+                    coords[d] += 2;
+                    auto&& right = apply(u_m, coords);
+
+                    res += hvector_m[d] * (left_boundary * left + not_right_boundary * right);
```

Comment:
Same here: add a comment to explain left_boundary*left and not_right_boundary*right

Reply:
I added comments :)